python/pyiceberg/expressions/__init__.py

```diff
@@ -48,12 +64,13 @@ class Bound(ABC):
     """Represents a bound value expression"""
 
 
-class Unbound(Generic[B], ABC):
+class Unbound(ABC):
     """Represents an unbound value expression"""
 
+    @property
     @abstractmethod
-    def bind(self, schema: Schema, case_sensitive: bool = True) -> B:
-        ...  # pragma: no cover
+    def as_bound(self) -> Type[Bound]:
```

Comment:
Is this a necessary change? I'd like to avoid extra changes if they aren't needed.

Reply:
I think so. Because we can use OOP now (because we've removed the data classes), we're able to solve this using inheritance in a cleaner way. Also, we remove all the types and return always a `BooleanExpression` after binding, which makes us lose the type information.

Also, we're not able to pass in the `[T]`:
```python
@dataclass(frozen=True)
class GreaterThan(LiteralPredicate[T]):
    as_bound = BoundGreaterThan[T]
```

Produces:
```
python/pyiceberg/expressions/__init__.py:518: error: Can't use bound type variable "T" to define generic alias
Found 1 error in 1 file (checked 83 source files)
```

But most importantly, we get no type checking at all. On current master, if I add some bogus arguments:

```python
@dataclass(frozen=True)
class UnaryPredicate(UnboundPredicate[T]):
    def bind(self, schema: Schema, case_sensitive: bool = True) -> BooleanExpression:
        bound_term = self.term.bind(schema, case_sensitive)
        return self.as_bound(bound_term, 1, 2, 3) # The int arguments are not expected
```
It just runs fine:
```
➜  python git:(master) ✗ pre-commit run --all-files mypy
mypy.....................................................................Passed
```

If I do the same on this branch:
```
➜  python git:(fd-remove-dataclass) pre-commit run --all-files mypy 
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

python/pyiceberg/expressions/__init__.py:325: error: Too many arguments for "BoundUnaryPredicate"
Found 1 error in 1 file (checked 83 source files)
```

Reply:
Sounds good to me.